cpp/src/arrow/scalar.cc

```diff
@@ -583,6 +608,17 @@ Result<std::shared_ptr<Scalar>> StructScalar::field(FieldRef ref) const {
   }
 }
 
+RunEndEncodedScalar::RunEndEncodedScalar(std::shared_ptr<Scalar> value,
+                                         std::shared_ptr<DataType> type)
+    : Scalar{std::move(type), value->is_valid}, value{std::move(value)} {}
```

Comment:
should we have a DCHECK here (e.g. that the value type is of the right type, that `type` is actually an REE type)?

Reply:
I didn't see this check in other scalars, so I assumed this is supposed to be handled in another layer.

Reply:
Hmm, I don't think it is - basically it's here or in validate

Reply:
I looked at other Scalars now and they check indirectly by performing some checked casts, so I will add a check here too. :+1: